app/main.js

```diff
@@ -14,57 +15,63 @@ let noteSectionsCustomization = {};
-        alg: "RS256",
-        typ: "JWT",
-    };
-    const payload = {
-        sub: OAUTH_CLIENT_UUID,
-        iss: OAUTH_CLIENT_UUID,
-        aud: `https://${REGION}.api.nabla.com/v1/core/server/oauth/token`,
-        exp: nowSeconds + 60,
-        iat: nowSeconds,
-    };
-    const jwtAssertion = KJUR.jws.JWS.sign(assertionHeader.alg, JSON.stringify(assertionHeader), JSON.stringify(payload), OAUTH_CLIENT_PRIVATE_KEY);
-    const response = await fetch(`https://${REGION}.api.nabla.com/v1/core/server/oauth/token`, {
-        method: 'POST',
-        headers: {
-            'Content-Type': 'application/json',
-        },
-        body: JSON.stringify({
-            grant_type: "client_credentials",
-            client_assertion_type: "urn:ietf:params:oauth:client-assertion-type:jwt-bearer",
-            client_assertion: jwtAssertion,
-        })
-    });
+let userAccessToken = INITIAL_USER_ACCESS_TOKEN;
+let userRefreshToken = INITIAL_USER_REFRESH_TOKEN;
 
-    return response.json();
-};
+const showTokenError = (message) => {
+    const errorDiv = document.getElementById("token-error");
+    if (!errorDiv) return;
+    errorDiv.innerHTML = message;
+    errorDiv.classList.remove("hide");
+}
 
-let serverAccessTokenCache = {
-    accessToken: null,
-    expiresAt: 0,
-};
+const decodeJWT = (token) => {
+    const parts = token.split('.');
+    if (parts.length !== 3) {
+        showTokenError("The user tokens seem invalid. You maybe forgot to provide initial tokens in the source code.");
+        throw new Error("Invalid JWT token");
+    }
+    const payload = parts[1];
+    return JSON.parse(atob(payload.replace(/-/g, '+').replace(/_/g, '/'))); // replace URL-safe characters
+}
 
-const getOrRefetchServerAccessToken = async () => {
+const isTokenExpiredOrExpiringSoon = (token) => {
     const nowSeconds = Math.floor(Date.now() / 1000);
+    return (decodeJWT(token).exp - nowSeconds) < 5;
+}
 
-    if (
-        serverAccessTokenCache.accessToken &&
-        nowSeconds < (serverAccessTokenCache.expiresAt - 5)
-    ) {
-        return serverAccessTokenCache.accessToken;
+const setUserTokens = (newAccessToken, newRefreshToken) => {
+    userAccessToken = newAccessToken;
+    userRefreshToken = newRefreshToken;
+}
+
+const getOrRefetchUserAccessToken = async () => {
+    if (!isTokenExpiredOrExpiringSoon(userAccessToken)) {
+        return userAccessToken;
     }
 
-    const { access_token, expires_in } = await fetchServerAccessToken();
-    serverAccessTokenCache.accessToken = access_token;
-    serverAccessTokenCache.expiresAt = nowSeconds + expires_in;
-    return access_token;
-};
+    if (isTokenExpiredOrExpiringSoon(userRefreshToken)) {
+        showTokenError("Your user refresh token has expired. Please provide new initial tokens in the source code.");
+        throw new Error("Refresh token expired");
+    }
+
+    const refreshResponse = await fetch(`https://${CORE_API_HOST}/v1/core/user/jwt/refresh`, {
```

Comment:
very nit but for clarity I'd share a ```const CORE_API_BASE_URL = `${REGION}.api.nabla.com/v1/core` ``` instead of `CORE_API_HOST` to only append `/server/xxx` or `/user/xxx` later when used